File: rb/test/discovery_entity_test.rb
# Discovery entity test

require "minitest/autorun"
require "json"
require_relative "../WeUltrarich_sdk"
require_relative "runner"

class DiscoveryEntityTest < Minitest::Test
  def test_create_instance
    testsdk = WeUltrarichSDK.test(nil, nil)
    ent = testsdk.Discovery(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "discovery" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = WeUltrarichSDK.test(seed, nil)
    seen = base.Discovery(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = WeUltrarichConfig.make_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = WeUltrarichSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.Discovery(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = discovery_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "discovery." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set WEULTRARICH_TEST_DISCOVERY_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    discovery_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.discovery")))
    discovery_ref01_data = nil
    if discovery_ref01_data_raw.length > 0
      discovery_ref01_data = Helpers.to_map(discovery_ref01_data_raw[0][1])
    end

    # LIST
    discovery_ref01_ent = client.Discovery(nil)
    discovery_ref01_match = {}

    discovery_ref01_list_result = discovery_ref01_ent.list(discovery_ref01_match, nil)
    assert discovery_ref01_list_result.is_a?(Array)

  end
end

def discovery_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "discovery", "DiscoveryTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = WeUltrarichSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["discovery01", "discovery02", "discovery03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["WEULTRARICH_TEST_DISCOVERY_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "WEULTRARICH_TEST_DISCOVERY_ENTID" => idmap,
    "WEULTRARICH_TEST_LIVE" => "FALSE",
    "WEULTRARICH_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["WEULTRARICH_TEST_DISCOVERY_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["WEULTRARICH_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = WeUltrarichSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["WEULTRARICH_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["WEULTRARICH_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: ts/test/entity/discovery/DiscoveryEntity.test.ts


import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'
import { createLiveTransport } from '../../live-runner'
import { runLiveEntity } from '../../live-entity'


import { WeUltrarichSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveClientOptions,
  liveDelay,
  loadEnvLocal,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


// AFTER the imports on purpose: TypeScript hoists `import` above any
// statement in the emitted CommonJS, so a loader placed above them would
// run only after every imported module had already been evaluated - and
// anything reading process.env at module scope would miss these values.
loadEnvLocal(__dirname + '/../../../.env.local')


describe('DiscoveryEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when WE_ULTRARICH_TEST_LIVE=TRUE.
  afterEach(liveDelay('WE_ULTRARICH_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = WeUltrarichSDK.test()
    const ent = testsdk.Discovery()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.WE_ULTRARICH_TEST_LIVE
    for (const op of ['list']) {
      if (!live && maybeSkipControl(t, 'entityOp', 'discovery.' + op, live)) return
    }

    
    const setup = basicSetup()
    if (setup.live) {
      return runLiveEntity(setup, {"active":true,"alias":{"field":{}},"fields":[{"active":true,"name":"route","req":true,"short":"The path to each available wealth expression.","type":"`$STRING`","index$":0}],"name":"discovery","op":{"list":{"input":"data","name":"list","points":[{"active":true,"args":{},"contract":{"id":"GET /expressions","json":"{\"operationId\":\"expressions\",\"parameters\":[],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"data\":{\"example\":[{\"route\":\"/v1/durationOfDailySpend\"},{\"route\":\"/v1/heightOfMoneyStack\"},{\"route\":\"/v1/numberOfItems\"},{\"route\":\"/v1/growthOfCompoundInterest\"},{\"route\":\"/v1/comparison\"}],\"items\":{\"properties\":{\"route\":{\"description\":\"The path to each available wealth expression.\",\"type\":\"string\"}},\"required\":[\"route\"],\"type\":\"object\"},\"type\":\"array\"},\"status\":{\"example\":\"success\",\"type\":\"string\"}},\"required\":[\"status\",\"data\"],\"type\":\"object\"}}},\"description\":\"a JSON object with response data\"},\"429\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"error\":{\"properties\":{\"code\":{\"example\":429,\"type\":\"integer\"},\"details\":{\"example\":\"Rate limit exceeded. Try again in 60 seconds.\",\"type\":\"string\"},\"message\":{\"example\":\"Too Many Requests\",\"type\":\"string\"},\"retryAfter\":{\"description\":\"The number of seconds remaining until the rate-limit window resets.\",\"example\":60,\"type\":\"integer\"}},\"required\":[\"code\",\"message\",\"details\",\"retryAfter\"],\"type\":\"object\"},\"status\":{\"example\":\"error\",\"type\":\"string\"}},\"required\":[\"status\",\"error\"],\"type\":\"object\"}}},\"description\":\"rate limit exceeded\",\"headers\":{\"Retry-After\":{\"description\":\"The number of seconds remaining until the rate-limit window resets.\",\"schema\":{\"example\":60,\"type\":\"integer\"}}}}},\"security\":[],\"securitySource\":\"definition\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/expressions","segments":[{"lit":"expressions"}],"select":{},"transform":{"req":"`reqdata`","res":"`body.data`"},"index$":0}],"key$":"list"}},"relations":{"ancestors":[]},"key$":"discovery","name__orig":"discovery","Name":"Discovery","name_":"discovery","name-":"discovery","NAME":"DISCOVERY","index$":1}, {"active":true,"entity":"discovery","key$":"BasicDiscoveryFlow","kind":"basic","name":"BasicDiscoveryFlow","param":{},"step":[{"active":true,"data":{},"input":{},"match":{},"op":"list","spec":[],"valid":[{"apply":"ItemExists","def":{"ref":"discovery_ref01"}}],"index$":0}]}, 'Discovery')
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let discovery_ref01_data = Object.values(setup.data.existing.discovery)[0] as any

    // LIST
    const discovery_ref01_ent = client.Discovery()
    const discovery_ref01_match: any = {}

    const discovery_ref01_list = (await discovery_ref01_ent.list(discovery_ref01_match)).map((e: any) => e.data())


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/discovery/DiscoveryTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = WeUltrarichSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['discovery01','discovery02','discovery03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'WE_ULTRARICH_TEST_DISCOVERY_ENTID': idmap,
    'WE_ULTRARICH_TEST_LIVE': 'FALSE',
    'WE_ULTRARICH_TEST_EXPLAIN': 'FALSE',
  })

  idmap = env['WE_ULTRARICH_TEST_DISCOVERY_ENTID']

  const live = 'TRUE' === env.WE_ULTRARICH_TEST_LIVE

  const transport = createLiveTransport()
  if (live) {
    const rawIds = process.env['WE_ULTRARICH_TEST_DISCOVERY_ENTID']
    idmap = rawIds && rawIds.trim() ? JSON.parse(rawIds) : {}
    if (!idmap || Array.isArray(idmap) || typeof idmap !== 'object') {
      throw new Error('Live ENTID must be a JSON object')
    }
    client = new WeUltrarichSDK(merge([
      // FIRST, so the generated fields below win: sdk-test-control.json's
      // test.client.options adds to the live client, it does not redirect it.
      liveClientOptions(),
      {
      },
      // 'extra || {}', not a bare 'extra': struct.merge returns UNDEFINED when the
      // last entry is undefined, and basicSetup is normally called with no
      // argument at all - so a bare 'extra' silently discarded the apikey
      // and server values above and handed the SDK undefined. Harmless
      // while there was nothing in that object; not harmless now.
      extra || {},
      { system: { fetch: transport.fetch } }
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.WE_ULTRARICH_TEST_EXPLAIN,
    live,
    transport,
    now: Date.now(),
  }

  return setup
}
  
